pushsync/pusher.go

```diff
@@ -0,0 +1,350 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pushsync
+
+import (
+	"context"
+	"encoding/hex"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/spancontext"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/opentracing/opentracing-go"
+	olog "github.com/opentracing/opentracing-go/log"
+)
+
+// DB interface implemented by localstore
+type DB interface {
+	// subscribe to chunk to be push synced - iterates from earliest to newest
+	SubscribePush(context.Context) (<-chan storage.Chunk, func())
+	// called to set a chunk as synced - and allow it to be garbage collected
+	// TODO this should take ... last argument to delete many in one batch
+	Set(context.Context, chunk.ModeSet, ...storage.Address) error
+}
+
+// Pusher takes care of the push syncing
+type Pusher struct {
+	store         DB                     // localstore DB
+	tags          *chunk.Tags            // tags to update counts
+	quit          chan struct{}          // channel to signal quitting on all loops
+	closed        chan struct{}          // channel to signal sync loop terminated
+	pushed        map[string]*pushedItem // cache of items push-synced
+	receipts      chan []byte            // channel to receive receipts
+	ps            PubSub                 // PubSub interface to send chunks and receive receipts
+	logger        log.Logger             // custom logger
+	retryInterval time.Duration          // dynamically adjusted time interval between retriess
+}
+
+const maxMeasurements = 20000
+
+// pushedItem captures the info needed for the pusher about a chunk during the
+// push-sync--receipt roundtrip
+type pushedItem struct {
+	tag         *chunk.Tag       // tag for the chunk
+	shortcut    bool             // if the chunk receipt was sent by self
+	firstSentAt time.Time        // first sent at time
+	lastSentAt  time.Time        // most recently sent at time
+	synced      bool             // set when chunk got synced
+	span        opentracing.Span // roundtrip span
+}
+
+// NewPusher contructs a Pusher and starts up the push sync protocol
+// takes
+// - a DB interface to subscribe to push sync index to allow iterating over recently stored chunks
+// - a pubsub interface to send chunks and receive statements of custody
+// - tags that hold the tags
+func NewPusher(store DB, ps PubSub, tags *chunk.Tags) *Pusher {
+	p := &Pusher{
+		store:         store,
+		tags:          tags,
+		quit:          make(chan struct{}),
+		closed:        make(chan struct{}),
+		pushed:        make(map[string]*pushedItem),
+		receipts:      make(chan []byte),
+		ps:            ps,
+		logger:        log.New("self", hex.EncodeToString(ps.BaseAddr())),
```

Comment:
please add `[:16]` after `hex.EncodeToString` so that the output is not too long